diff --git a/arch/mips/bits/termios.h b/arch/mips/bits/termios.h
index f559f76..6a1205d 100644
--- a/arch/mips/bits/termios.h
+++ b/arch/mips/bits/termios.h
@@ -141,9 +141,9 @@ struct termios {
 #define TCOFLUSH  1
 #define TCIOFLUSH 2
 
-#define TCSANOW 0x540e
-#define TCSADRAIN 0x540f
-#define TCSAFLUSH 0x5410
+#define TCSANOW   0
+#define TCSADRAIN 1
+#define TCSAFLUSH 2
 
 #if defined(_GNU_SOURCE) || defined(_BSD_SOURCE)
 #define EXTA    0000016
diff --git a/arch/mips64/bits/termios.h b/arch/mips64/bits/termios.h
index f559f76..6a1205d 100644
--- a/arch/mips64/bits/termios.h
+++ b/arch/mips64/bits/termios.h
@@ -141,9 +141,9 @@ struct termios {
 #define TCOFLUSH  1
 #define TCIOFLUSH 2
 
-#define TCSANOW 0x540e
-#define TCSADRAIN 0x540f
-#define TCSAFLUSH 0x5410
+#define TCSANOW   0
+#define TCSADRAIN 1
+#define TCSAFLUSH 2
 
 #if defined(_GNU_SOURCE) || defined(_BSD_SOURCE)
 #define EXTA    0000016
diff --git a/arch/mipsn32/bits/termios.h b/arch/mipsn32/bits/termios.h
index f559f76..6a1205d 100644
--- a/arch/mipsn32/bits/termios.h
+++ b/arch/mipsn32/bits/termios.h
@@ -141,9 +141,9 @@ struct termios {
 #define TCOFLUSH  1
 #define TCIOFLUSH 2
 
-#define TCSANOW 0x540e
-#define TCSADRAIN 0x540f
-#define TCSAFLUSH 0x5410
+#define TCSANOW   0
+#define TCSADRAIN 1
+#define TCSAFLUSH 2
 
 #if defined(_GNU_SOURCE) || defined(_BSD_SOURCE)
 #define EXTA    0000016